In [6]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os

pd.set_option("display.max_columns", None)

In [7]:
prediction_df = pd.read_csv("../../05_outputs/predictions/churn_predictions.csv")

prediction_df.head()

,customer_id,churn_probability,predicted_churn,risk_level
0,50001,0.705,1,High Risk
1,50002,0.955,1,High Risk
2,50003,0.985,1,High Risk
3,50004,0.910,1,High Risk
4,50005,0.900,1,High Risk


In [8]:
import sys

sys.path.append(os.path.abspath("../preprocessing"))

from preprocess_churn import load_gold_data

gold_df = load_gold_data()

gold_df.head()

,customer_id,churn,tenure,city_tier,warehouse_to_home,hour_spend_on_app,number_of_device_registered,satisfaction_score,number_of_address,complain,order_amount_hike_from_last_year,coupon_used,order_count,day_since_last_order,cashback_amount,preferred_login_device,preferred_payment_mode,gender,preferred_order_cat,marital_status,tenure_missing_flag,warehouse_to_home_missing_flag,hour_spend_on_app_missing_flag,day_since_last_order_missing_flag,is_new_customer,low_satisfaction_flag,has_complaint,inactive_customer_flag,high_cashback_customer_flag,batch_id,loaded_at,loaded_by
0,50001,1,4.0,3,6.0,3.0,3.0,2,9.0,1,11.0,1.0,1.0,5.0,160.0,mobile phone,debit card,female,laptop & accessory,single,0,0,0,0,1,1,1,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user
1,50002,1,0.0,1,8.0,3.0,4.0,3,7.0,1,15.0,0.0,1.0,0.0,121.0,phone,upi,male,mobile,single,1,0,0,0,1,0,1,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user
2,50003,1,0.0,1,30.0,2.0,4.0,3,6.0,1,14.0,0.0,1.0,3.0,120.0,phone,debit card,male,mobile,single,1,0,0,0,1,0,1,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user
3,50004,1,0.0,3,15.0,2.0,4.0,5,8.0,0,23.0,0.0,1.0,3.0,134.0,phone,debit card,male,laptop & accessory,single,0,0,0,0,1,0,0,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user
4,50005,1,0.0,1,12.0,0.0,3.0,5,3.0,0,11.0,1.0,1.0,3.0,130.0,phone,cc,male,mobile,single,0,0,1,0,1,0,0,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user


In [9]:
segmentation_df = prediction_df.merge(
    gold_df,
    on="customer_id",
    how="left"
)

segmentation_df.head()

,customer_id,churn_probability,predicted_churn,risk_level,churn,tenure,city_tier,warehouse_to_home,hour_spend_on_app,number_of_device_registered,satisfaction_score,number_of_address,complain,order_amount_hike_from_last_year,coupon_used,order_count,day_since_last_order,cashback_amount,preferred_login_device,preferred_payment_mode,gender,preferred_order_cat,marital_status,tenure_missing_flag,warehouse_to_home_missing_flag,hour_spend_on_app_missing_flag,day_since_last_order_missing_flag,is_new_customer,low_satisfaction_flag,has_complaint,inactive_customer_flag,high_cashback_customer_flag,batch_id,loaded_at,loaded_by
0,50001,0.705,1,High Risk,1,4.0,3,6.0,3.0,3.0,2,9.0,1,11.0,1.0,1.0,5.0,160.0,mobile phone,debit card,female,laptop & accessory,single,0,0,0,0,1,1,1,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user
1,50002,0.955,1,High Risk,1,0.0,1,8.0,3.0,4.0,3,7.0,1,15.0,0.0,1.0,0.0,121.0,phone,upi,male,mobile,single,1,0,0,0,1,0,1,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user
2,50003,0.985,1,High Risk,1,0.0,1,30.0,2.0,4.0,3,6.0,1,14.0,0.0,1.0,3.0,120.0,phone,debit card,male,mobile,single,1,0,0,0,1,0,1,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user
3,50004,0.910,1,High Risk,1,0.0,3,15.0,2.0,4.0,5,8.0,0,23.0,0.0,1.0,3.0,134.0,phone,debit card,male,laptop & accessory,single,0,0,0,0,1,0,0,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user
4,50005,0.900,1,High Risk,1,0.0,1,12.0,0.0,3.0,5,3.0,0,11.0,1.0,1.0,3.0,130.0,phone,cc,male,mobile,single,0,0,1,0,1,0,0,0,0,f190cdea-1610-40b4-8591-813dff94068a,2026-06-17 14:50:23.631095,local_user


In [10]:
segmentation_df.shape

(5630, 35)

In [11]:
segmentation_df["risk_level"].value_counts()

risk_level
Low Risk       4680
High Risk       853
Medium Risk      97
Name: count, dtype: int64

In [12]:
segmentation_df["risk_level"].value_counts(normalize=True) * 100

risk_level
Low Risk       83.126110
High Risk      15.150977
Medium Risk     1.722913
Name: proportion, dtype: float64

In [13]:
high_risk_df = segmentation_df[
    segmentation_df["risk_level"] == "High Risk"
].copy()

high_risk_df.shape

(853, 35)

In [14]:
high_risk_df["complain"].value_counts(normalize=True) * 100

complain
1    55.685815
0    44.314185
Name: proportion, dtype: float64

In [15]:
high_risk_df["tenure"].describe()

count    853.000000
mean       2.631887
std        4.828808
min        0.000000
25%        0.000000
50%        1.000000
75%        1.000000
max       21.000000
Name: tenure, dtype: float64

In [16]:
high_risk_df["preferred_order_cat"].value_counts()

preferred_order_cat
mobile phone          327
mobile                198
laptop & accessory    185
fashion               118
grocery                19
others                  6
Name: count, dtype: int64

In [17]:
segmentation_df["preferred_payment_mode"] = segmentation_df["preferred_payment_mode"].replace({
    "cod": "cash on delivery",
    "cc": "credit card"
})

segmentation_df["preferred_login_device"] = segmentation_df["preferred_login_device"].replace({
    "phone": "mobile phone"
})

In [18]:
high_risk_df = segmentation_df[
    segmentation_df["risk_level"] == "High Risk"
].copy()

In [19]:
high_risk_df["preferred_payment_mode"].value_counts()

preferred_payment_mode
debit card          325
credit card         221
e wallet            133
cash on delivery    106
upi                  68
Name: count, dtype: int64

In [20]:
high_risk_df["preferred_payment_mode"].value_counts(normalize=True) * 100

preferred_payment_mode
debit card          38.100821
credit card         25.908558
e wallet            15.592028
cash on delivery    12.426729
upi                  7.971864
Name: proportion, dtype: float64

In [21]:
risk_numeric_summary = segmentation_df.groupby("risk_level")[
    [
        "tenure",
        "cashback_amount",
        "warehouse_to_home",
        "day_since_last_order",
        "order_count",
        "coupon_used",
        "satisfaction_score"
    ]
].mean()

risk_numeric_summary

,tenure,cashback_amount,warehouse_to_home,day_since_last_order,order_count,coupon_used,satisfaction_score
risk_level,,,,,,,
High Risk,2.631887,158.248535,15.798359,2.923798,2.682298,1.667057,3.406800
Low Risk,11.070726,180.725427,14.785043,4.549145,2.893590,1.667308,3.001496
Medium Risk,6.422680,175.010309,15.020619,4.134021,3.391753,1.907216,3.226804


In [22]:
os.makedirs("../../05_outputs/segments", exist_ok=True)

high_risk_df.to_csv(
    "../../05_outputs/segments/high_risk_customers.csv",
    index=False
)

os.path.exists("../../05_outputs/segments/high_risk_customers.csv")

True

In [23]:
risk_numeric_profile = (
    segmentation_df
    .groupby("risk_level")
    .agg(
        customers=("customer_id", "count"),
        churn_rate=("churn", "mean"),
        avg_churn_probability=("churn_probability", "mean"),
        avg_tenure=("tenure", "mean"),
        median_tenure=("tenure", "median"),
        complaint_rate=("complain", "mean"),
        has_complaint_rate=("has_complaint", "mean"),
        avg_days_since_last_order=("day_since_last_order", "mean"),
        avg_order_count=("order_count", "mean"),
        avg_coupon_used=("coupon_used", "mean"),
        avg_cashback_amount=("cashback_amount", "mean"),
        avg_satisfaction_score=("satisfaction_score", "mean"),
        avg_warehouse_to_home=("warehouse_to_home", "mean")
    )
    .reset_index()
)

risk_numeric_profile

,risk_level,customers,churn_rate,avg_churn_probability,avg_tenure,median_tenure,complaint_rate,has_complaint_rate,avg_days_since_last_order,avg_order_count,avg_coupon_used,avg_cashback_amount,avg_satisfaction_score,avg_warehouse_to_home
0,High Risk,853,1.000000,0.883453,2.631887,1.0,0.556858,0.556858,2.923798,2.682298,1.667057,158.248535,3.406800,15.798359
1,Low Risk,4680,0.001923,0.028261,11.070726,10.0,0.232906,0.232906,4.549145,2.893590,1.667308,180.725427,3.001496,14.785043
2,Medium Risk,97,0.886598,0.590619,6.422680,2.0,0.402062,0.402062,4.134021,3.391753,1.907216,175.010309,3.226804,15.020619


In [24]:
risk_category_profile = (
    segmentation_df
    .groupby(["risk_level", "preferred_order_cat"])
    .agg(customers=("customer_id", "count"))
    .reset_index()
)

risk_category_profile["share_within_risk_level"] = (
    risk_category_profile["customers"]
    / risk_category_profile.groupby("risk_level")["customers"].transform("sum")
)

risk_category_profile.sort_values(
    ["risk_level", "share_within_risk_level"],
    ascending=[True, False]
)

,risk_level,preferred_order_cat,customers,share_within_risk_level
4,High Risk,mobile phone,327,0.383353
3,High Risk,mobile,198,0.232122
2,High Risk,laptop & accessory,185,0.216882
0,High Risk,fashion,118,0.138335
1,High Risk,grocery,19,0.022274
5,High Risk,others,6,0.007034
8,Low Risk,laptop & accessory,1839,0.392949
10,Low Risk,mobile phone,917,0.195940
6,Low Risk,fashion,698,0.149145
9,Low Risk,mobile,590,0.126068


In [25]:
import os

os.makedirs("../../05_outputs/segments", exist_ok=True)

risk_numeric_profile.to_csv(
    "../../05_outputs/segments/risk_level_numeric_profile.csv",
    index=False
)

risk_category_profile.to_csv(
    "../../05_outputs/segments/risk_level_category_profile.csv",
    index=False
)

## Segmentation Insights

The prediction output was merged with the Gold customer table to analyze customer churn risk segments.

The model classified 853 customers as High Risk, representing 15.15% of the total customer base. Most customers were classified as Low Risk, while only a small share fell into the Medium Risk group.

The purpose of this segmentation is not only to label customers as High, Medium, or Low Risk, but also to understand how these groups differ and what actions the business should take for each group.

### High Risk Segment

High Risk customers show several clear data patterns.

First, this group has very low tenure. The average tenure of High Risk customers is around 2.63, compared with around 11.07 for Low Risk customers. The median tenure of the High Risk group is only 1. This suggests that many churn-risk customers are still in the early stage of their customer lifecycle.

Business implication:

Customers with short tenure may not have built strong loyalty or repeat-purchase habits yet. Therefore, the company should not wait until these customers become inactive. Early onboarding and engagement actions are important.

Recommended actions:

* Create onboarding campaigns for new or low-tenure customers.
* Send first-month or second-purchase incentives.
* Track whether new customers make another purchase within a short period.
* Provide personalized product recommendations after the first order.

Second, complaint behavior is strongly concentrated in the High Risk group. More than half of High Risk customers have submitted complaints. This suggests that service experience is an important churn driver.

Business implication:

For these customers, a generic discount may not be enough. If the churn signal comes from a complaint, the company should first fix the service issue before sending promotions.

Recommended actions:

* Prioritize High Risk customers with complaint history.
* Create a complaint recovery workflow.
* Follow up with affected customers after issue resolution.
* Offer compensation, cashback, or service recovery vouchers only after the complaint is addressed.

Third, High Risk customers are strongly concentrated in mobile-related categories. Mobile phone accounts for 38.34% of High Risk customers, and mobile accounts for 23.21%. Together, mobile-related categories represent 61.55% of the High Risk segment.

This is much higher than the Low Risk group, where mobile phone and mobile together represent around 32.20%.

Business implication:

Mobile-related purchases may be more one-off or replacement-driven, meaning customers may have fewer natural reasons to return after the first purchase. These categories may require a different retention strategy compared with repeat-driven categories such as grocery.

Recommended actions:

* Design mobile-category retention campaigns.
* Use accessory cross-sell campaigns after mobile purchases.
* Send warranty, care, or usage-related follow-up messages.
* Offer targeted cashback for second purchases in related categories.

Finally, High Risk customers also have lower average cashback amounts than Low Risk customers. This may suggest weaker incentive exposure or lower engagement with cashback-based retention mechanisms.

Business implication:

Targeted incentives may help increase engagement, especially when combined with low tenure and complaint signals.

Recommended actions:

* Use targeted cashback instead of broad discounts.
* Prioritize incentives for High Risk customers with low tenure, complaint history, or mobile-related purchases.
* Avoid giving large discounts to all customers equally.

### Medium Risk Segment

Medium Risk customers show mixed behavior. They are not as urgent as High Risk customers, but they already show some churn signals.

In terms of preferred order category, Medium Risk customers are spread across mobile phone, laptop & accessory, and mobile. This suggests that Medium Risk is a transition group rather than a single clearly defined customer type.

Business implication:

Medium Risk customers should be monitored before they become High Risk. The business should use lighter and lower-cost engagement actions for this group.

Recommended actions:

* Send light incentives such as coupons or free shipping.
* Monitor changes in days since last order, complaint status, and order frequency.
* Trigger stronger retention actions if the customer moves from Medium Risk to High Risk.
* Use personalized product recommendations instead of aggressive discounting.

### Low Risk Segment

Low Risk customers represent the majority of the customer base. This group is more concentrated in laptop & accessory, which accounts for 39.29% of Low Risk customers.

Compared with the High Risk group, Low Risk customers have higher tenure and lower complaint concentration. This suggests that they are relatively more stable and should not receive aggressive retention spending.

Business implication:

The company should maintain engagement with Low Risk customers but avoid unnecessary discounts.

Recommended actions:

* Maintain loyalty communication.
* Send regular product updates or personalized recommendations.
* Avoid over-discounting customers who are already stable.
* Continue monitoring for future changes in churn risk.

### Key Takeaway

The main insight from the risk segmentation is that High Risk customers are not just randomly selected by the model. They show clear behavioral patterns: lower tenure, higher complaint concentration, stronger concentration in mobile-related categories, and lower cashback amounts.

Therefore, retention actions should not be generic. The company should prioritize:

* Onboarding and early engagement for low-tenure customers.
* Complaint recovery for customers with service issues.
* Mobile-category cross-sell and second-purchase campaigns.
* Targeted cashback for selected High Risk customers.

The High Risk customer segment was saved to:

```text
05_outputs/segments/high_risk_customers.csv
```

The risk-level profiling outputs were saved to:

```text
05_outputs/segments/risk_level_numeric_profile.csv
05_outputs/segments/risk_level_category_profile.csv
```
